### ISSUE TITLE
Latest Mono produces Portable .pdb instead of .mdb(Mono.Cecil update required)

### ISSUE THREAD

**Issue by DavidKarlas** [nunit/nunit3-vs-adapter#327]
Latest Mono is producing Portable .pdb instead of .mdb by default, resulting in this exception:
```
 System.IO.FileNotFoundException: Could not find file "/Users/davidkarlas/Projects/nunit24214asf/nunit24214asf/bin/Debug/nunit24214asf.dll.mdb".
File name: '/Users/davidkarlas/Projects/nunit24214asf/nunit24214asf/bin/Debug/nunit24214asf.dll.mdb'
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00195] in /private/tmp/source-mono-2017-04/bockbuild-xamarin/profiles/mono-mac-xamarin/build-root/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:232 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean isAsync, System.Boolean anonymous) [0x00000] in /private/tmp/source-mono-2017-04/bockbuild-xamarin/profiles/mono-mac-xamarin/build-root/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:149 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access) [0x00000] in /private/tmp/source-mono-2017-04/bockbuild-xamarin/profiles/mono-mac-xamarin/build-root/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:86 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream:.ctor (string,System.IO.FileMode,System.IO.FileAccess)
  at Mono.CompilerServices.SymbolWriter.MonoSymbolFile.ReadSymbolFile (System.String mdbFilename) [0x00000] in <682b0e3580fe403dad53adbaea0f93fd>:0 
  at Mono.CompilerServices.SymbolWriter.MonoSymbolFile.ReadSymbolFile (System.String mdbFilename, System.Guid assemblyGuid) [0x00000] in <682b0e3580fe403dad53adbaea0f93fd>:0 
  at Mono.Cecil.Mdb.MdbReaderProvider.GetSymbolReader (Mono.Cecil.ModuleDefinition module, System.String fileName) [0x00012] in <682b0e3580fe403dad53adbaea0f93fd>:0 
  at Mono.Cecil.ModuleReader.ReadSymbols (Mono.Cecil.ModuleDefinition module, Mono.Cecil.ReaderParameters parameters) [0x00032] in <ce25e56d9faa4073b3cc14996c972301>:0 
  at Mono.Cecil.ModuleReader.CreateModuleFrom (Mono.Cecil.PE.Image image, Mono.Cecil.ReaderParameters parameters) [0x00042] in <ce25e56d9faa4073b3cc14996c972301>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.IO.Stream stream, Mono.Cecil.ReaderParameters parameters) [0x00028] in <ce25e56d9faa4073b3cc14996c972301>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x0000a] in <ce25e56d9faa4073b3cc14996c972301>:0 
  at NUnit.VisualStudio.TestAdapter.NavigationDataProvider.CacheNewTypes (System.String assemblyPath, System.Collections.Generic.IDictionary`2[TKey,TValue] types) [0x0003b] in <e1110a31ce52442d931c91db5aa14a6a>:0 
  at NUnit.VisualStudio.TestAdapter.NavigationDataProvider.CacheTypes (System.String assemblyPath) [0x00006] in <e1110a31ce52442d931c91db5aa14a6a>:0 
  at NUnit.VisualStudio.TestAdapter.NavigationDataProvider.GetNavigationData (System.String className, System.String methodName) [0x00015] in <e1110a31ce52442d931c91db5aa14a6a>:0 
  at NUnit.VisualStudio.TestAdapter.TestConverter.MakeTestCaseFromXmlNode (System.Xml.XmlNode testNode) [0x00058] in <e1110a31ce52442d931c91db5aa14a6a>:0 
  at NUnit.VisualStudio.TestAdapter.TestConverter.ConvertTestCase (System.Xml.XmlNode testNode) [0x0004c] in <e1110a31ce52442d931c91db5aa14a6a>:0 
  at NUnit.VisualStudio.TestAdapter.NUnit3TestDiscoverer.ProcessTestCases (System.Xml.XmlNode topNode, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.ITestCaseDiscoverySink discoverySink, NUnit.VisualStudio.TestAdapter.TestConverter testConverter) [0x00021] in <e1110a31ce52442d931c91db5aa14a6a>:0
```

Solution should be pretty simple, update to Mono.Cecil to 0.10 version(notice that it's still in beta NuGet feed) https://www.nuget.org/packages/Mono.Cecil/0.10.0-beta5 since Cecil autodetects and reads portable .pdb https://github.com/jbevain/cecil/blob/a2314c895987ee801a0c3e32dac05f5399beda37/Mono.Cecil.Cil/Symbols.cs#L726

Maybe also consider not failing to discover/run tests if .pdb/.mdb is not found.

**Comment by rprouse** [comment]
Thanks for the info. We are in a bit of a bind with Mono.Cecil 0.10 because it has dropped support for .NET 2.0 where our Engine is compiled 2.0 to allow running tests back to 2.0.

Carrying on if the PDB is a good stop-gap measure, but it will disable double clicking on your tests and navigating to the source in Visual Studio.

**Comment by rprouse** [comment]
@DavidKarlas which .NET version is your test project targeting and which version of Mono are you seeing this on?

**Comment by DavidKarlas** [comment]
Mono 5.0(went to stable yesterday) by default uses Roslyn compiler which produces Portable .pdb.

My unit test project is default Full Framework `v4.6.1` version.

I ran into this while implementing https://github.com/Microsoft/vstest into MonoDevelop(Visual Studio for Mac)

**Comment by rprouse** [comment]
@DavidKarlas thanks for the heads up. Supporting the newest compilers and frameworks is starting to conflict with supporting older frameworks, so we are going to have to figure out what to do as a team.

**Comment by rprouse** [comment]
@DavidKarlas if you want to follow, I am going to start discussion about fixes for this in #296 as it is closely related.

**Comment by CharliePoole** [comment]
It looks like the part about not crashing when we can't get the location info was fixed by Terje in #276. But I'm not clear if that fix applies to the newer .NET Core build.

@rprouse Maybe it's just me, but having the two distinct runners bundled in one package is making it quite difficult to figure out which one has a particular error. For example, we can't tell users to uninstall just one of them to see what happens.

**Comment by rprouse** [comment]
@CharliePoole to determine which runner is causing a problem, you just need to know what framework they are targeting. If it is .NET Framework, it is the original runner, if it is .NET Core, it is the .NET Core runner.

#276 is in 3.7 and 3.8-alpha-1, @DavidKarlas are you using either of those?

**Comment by CharliePoole** [comment]
@rprouse I understand, but that's a somewhat subtle point for users who often don't provide the info. (I realize the OP did in this case)

I'm also not sure it's a great idea to have two target runtimes in a package that invoke different codebases and have different capabilities. It makes it very hard to do updates of just one of them and the newer code will typically require more frequent update.

I'm the one who originally decided that the adapter should use the engine - there __was__ an alternative option discussed - and I'm beginning to think it wasn't such a great idea. I'm toying with the idea of creating an alternate version that bypasses the engine completely.

**Comment by DavidKarlas** [comment]
Seems to me like I'm using `3.8-alpha-1` and still seeing this... Btw, would be nice to also have this fixed for `NUnitTestAdapter`(v2)

**Comment by DavidKarlas** [comment]
Also that fix looks interesting since it would kinda mean to never load `.mdb` even if it exists.

**Comment by DavidKarlas** [comment]
Heh, what probably happens is... new Mono creates .pdb file, hence that PR sets `ReadSymbols = CheckIfPdbExists(assemblyName);` but after that Cecil tries to read `.mdb` since it's running on *nix+Mono and crashes...

**Comment by rprouse** [comment]
>I'm also not sure it's a great idea to have two target runtimes in a package that invoke different codebases and have different capabilities. It makes it very hard to do updates of just one of them and the newer code will typically require more frequent update.

The codebases are not that different. The engine that the .NET Core version uses is nearly identical to the codepaths through the full engine and the adapter code is nearly identical. The adapter doesn't use any of the features that differ in the two versions of the engine.

Also, if we add UWP support to the adapter, which should be very easy now, I wouldn't want to be maintaining multiple adapters for each target. It is not how MSTest or xUnit do it. They do it as we are now with 3.8. I would also not want our users to have to install multiple adapters because their project multi-targets frameworks.

**Comment by DavidKarlas** [comment]
Maybe you can add Issue Template which then looks something like this https://github.com/Microsoft/vstest/issues/new which specifically asks for target Framework...

**Comment by rprouse** [comment]
@DavidKarlas that is a good idea. We always ask questions like that, so we should prompt people when they create issues.

**Comment by DavidKarlas** [comment]
Any progress on this?

**Comment by rprouse** [comment]
@DavidKarlas sorry, no progress on this although the workaround works. It is fixed for .NET Core projects. I tried several different approaches, but the root issue is that the NUnit engine uses an older version of Mono.Cecil that targets .NET 2.0 so we can continue to run older tests under their target framework. Unfortunately, the version of Mono.Cecil that fixes this no longer targets .NET 2.0.

We as a team are going to have to decide how we are going to tackle this. One option would be to update the engine to .NET 3.5, but that would drop support for running tests under .NET 2.0. We could still run .NET 2.0 tests, but they would run under 3.5.

Another possible option is to start using DiaSession again in the adapter. I tried a branch with this without much luck.

@nunit/core-team and @nunit/engine-team what are your thoughts on this? I think that this is at root an Engine issue since it is the engine that is putting the constraints on the adapter.

**Comment by jnm2** [comment]
My opinion is that it's worth the work to get away from Mono.Cecil, whether that's DiaSession or Roslyn's metadata readers.
Ideally we offload as much of the work as possible to VS. Not knowing anything, it would seem that DiaSession is more likely to integrate in such a way that VS is doing as much of the work as possible.

I'd also like to put pressure on the VS IDE team to provide an interface which doesn't make us responsible for anything that isn't our domain. It makes no sense to me for us or any test framework to be responsible for resolving file and line from member symbols.

**Comment by codito** [comment]
@jnm2 in the post-15.3 versions of VS, we will be using the roslyn based APIs within VS to find file and line number. We're working on a vstest RFC with more details. Post this change, the adapters need not lookup for PDB information.

However for customers using older versions of VS, adapter still needs to support file/line number information. @rprouse we can help out if you can point to the issues you found with DiaSession.

**Comment by ChrisMaddock** [comment]
Regardless of possible paths forwards, my feeling is that building the engine as .NET 2.0 is a concern for a couple of big issues right now (this, x-plat agent/engine communication.)

I wonder if it's time to have a discussion about supporting .NET 2.0 in the same fashion as we did .NET 1.1 in NUnit 2 - i.e. we can run tests in .NET 2.0 via an agent, but the engine must be ran via .NET 3.5. NUnitLite of course remains an option for running purely in .NET 2.0, and I don't think we yet see the same issues with supporting .NET 2.0 in the framework.

**Comment by jnm2** [comment]
@codito That sounds great! Where is the discussion happening right now?

**Comment by codito** [comment]
@jnm2 here's the doc: https://github.com/Microsoft/vstest-docs/pull/59/files. Please share your feedback.

Is it possible to turn off source information collection in the nunit adapter? We can try and see the perf benefits.

**Comment by OsirisTerje** [comment]
@codito Currently no.  Could be added as a runsettings option though.....

**Comment by codito** [comment]
> Could be added as a runsettings option though.....

Here's the PR(https://github.com/nunit/nunit3-vs-adapter/pull/350) with `CollectSourceInformation` runsettings to disable collection of filename/linenumber in the adapter. It implements the RFC in [vstest-docs#59](https://github.com/Microsoft/vstest-docs/pull/59). @rprouse @OsirisTerje can you please review?

**Comment by jnm2** [comment]
This was fixed in #454 and is available now at https://www.nuget.org/packages/NUnit3TestAdapter/.
